File: src/main/java/com/jozufozu/exnihiloomnia/common/registries/recipes/FluidMixingRecipe.java
package com.jozufozu.exnihiloomnia.common.registries.recipes;

import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSyntaxException;
import com.jozufozu.exnihiloomnia.common.lib.LibRegistries;
import com.jozufozu.exnihiloomnia.common.registries.JsonHelper;
import net.minecraft.block.Block;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import net.minecraft.util.JsonUtils;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.SoundEvent;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.registries.IForgeRegistryEntry;

public class FluidMixingRecipe extends IForgeRegistryEntry.Impl<FluidMixingRecipe>
{
    /**
     * What's in the barrel
     */
    private FluidStack lower;
    
    /**
     * What goes on the barrel
     */
    private FluidStack upper;
    
    private ItemStack output;
    
    private SoundEvent craftSound;
    
    public FluidStack getLower()
    {
        return lower;
    }
    
    public FluidStack getUpper()
    {
        return upper;
    }
    
    public ItemStack getOutput()
    {
        return output.copy();
    }
    
    public SoundEvent getCraftSound()
    {
        return craftSound;
    }
    
    public boolean matches(FluidStack lower, FluidStack upper)
    {
        return this.upper.isFluidEqual(upper) && this.lower.isFluidEqual(lower);
    }

    public static FluidMixingRecipe deserialize(JsonObject object) throws JsonParseException
    {
        if (!object.has(LibRegistries.OUTPUT_BLOCK))
        {
            throw new JsonSyntaxException("Fluid mixing recipe is missing output!");
        }
    
        FluidMixingRecipe out = new FluidMixingRecipe();
        
        out.lower = new FluidStack(JsonHelper.deserializeFluid(object, LibRegistries.IN_BARREL), 1000);
        out.upper = new FluidStack(JsonHelper.deserializeFluid(object, LibRegistries.ON_BARREL), 1000);
        
        out.output = JsonHelper.deserializeItem(object.getAsJsonObject(LibRegistries.OUTPUT_BLOCK), true);
        
        out.craftSound = SoundEvent.REGISTRY.getObject(new ResourceLocation(JsonUtils.getString(object, "sound", "minecraft:block.lava.extinguish")));
        
        if (Block.getBlockFromItem(out.output.getItem()) == Blocks.AIR)
        {
            throw new JsonSyntaxException("Recipe does not output a block, ignoring");
        }
        
        return out;
    }
}

File: src/main/kotlin/com/jozufozu/exnihiloomnia/common/registries/recipes/FluidMixingRecipe.kt
package com.jozufozu.exnihiloomnia.common.registries.recipes

import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonSyntaxException
import com.jozufozu.exnihiloomnia.common.lib.LibRegistries
import com.jozufozu.exnihiloomnia.common.registries.JsonHelper
import com.jozufozu.exnihiloomnia.common.registries.RegistryLoader
import com.jozufozu.exnihiloomnia.common.util.contains
import net.minecraft.item.ItemStack
import net.minecraft.util.JSONUtils
import net.minecraft.util.ResourceLocation
import net.minecraft.util.SoundEvent
import net.minecraft.util.SoundEvents
import net.minecraftforge.common.crafting.CraftingHelper
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.registries.ForgeRegistries
import net.minecraftforge.registries.ForgeRegistryEntry

class FluidMixingRecipe(
        /**
         * What's in the barrel
         */
        val lower: FluidStack,

        /**
         * What goes on the barrel
         */
        val upper: FluidStack,

        output: ItemStack,

        val craftSound: SoundEvent = SoundEvents.BLOCK_LAVA_EXTINGUISH
): ForgeRegistryEntry<FluidMixingRecipe>() {

    val output: ItemStack = output
        get() = field.copy()

    fun matches(lower: FluidStack, upper: FluidStack): Boolean {
        return this.upper.isFluidEqual(upper) && this.lower.isFluidEqual(lower)
    }

    companion object Serde {

        private const val DEFAULT_SOUND: String = "minecraft:block.lava.extinguish"

        @Throws(JsonParseException::class)
        fun deserialize(recipe: JsonObject): FluidMixingRecipe? {
            if (LibRegistries.OUTPUT !in recipe) throw JsonSyntaxException("fluid mixing recipe is missing \"${LibRegistries.OUTPUT}\"")
            if (LibRegistries.IN_BARREL !in recipe) throw JsonSyntaxException("fluid mixing recipe is missing \"${LibRegistries.IN_BARREL}\"")
            if (LibRegistries.ON_BARREL !in recipe) throw JsonSyntaxException("fluid mixing recipe is missing \"${LibRegistries.ON_BARREL}\"")

            if (!CraftingHelper.processConditions(recipe, LibRegistries.CONDITIONS)) return null

            val lower = JsonHelper.deserializeFluid(recipe[LibRegistries.IN_BARREL])
            val upper = JsonHelper.deserializeFluid(recipe[LibRegistries.ON_BARREL])

            val output = JsonHelper.deserializeItem(JSONUtils.getJsonObject(recipe, LibRegistries.OUTPUT), true)

            val soundName = JSONUtils.getString(recipe, "sound", DEFAULT_SOUND)
            val craftSound = ForgeRegistries.SOUND_EVENTS.getValue(ResourceLocation(soundName))

            if (craftSound == null) {
                RegistryLoader.warn("'$soundName' is not a valid sound, defaulting to '$DEFAULT_SOUND'")
                return FluidMixingRecipe(lower, upper, output)
            }

            return FluidMixingRecipe(lower, upper, output, craftSound)
        }
    }
}
